### PR TITLE
Implementing LeashData

### PR DESCRIPTION
**Pull request by Rain336**
Implementation of `SpongeLeashData`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `ImmutableSpongeLeashData`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `LeashDualProcessor`:
- [x] Appropriately extended `AbstractSingleTargetDualProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `LEASH_HOLDER` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `DataRegistrar`
- [x] Registered the `ValueProcessor`s in the `DataRegistrar`

Test Plugin code:
```java
@Plugin(id = "CodeTester", name = "CodeTester", version = "1.0.0")
public class CodeTester {

    @Listener
    public void preInit(GamePreInitializationEvent event) {
        CommandSpec makeEntity = CommandSpec.builder()
                .arguments(GenericArguments.onlyOne(GenericArguments.catalogedElement(Texts.of("EntityType"), EntityType.class)))
                .executor((src, args) -> {
                    EntityType type = args.<EntityType>getOne("EntityType").get();
                    if(src instanceof Player) {
                        Player player = (Player) src;
                        Optional<Entity> optional = player.getLocation().getExtent().createEntity(type,
                                player.getLocation().getPosition().add(2, 0, 2));
                        if(!optional.isPresent()) {
                            player.sendMessage(Texts.of(TextColors.RED, "Error: Couldn't create Entity"));
                            return CommandResult.empty();
                        }
                        Entity entity = optional.get();
                        player.getLocation().getExtent().spawnEntity(entity, Cause.of(this));
                        if(entity.supports(Keys.LEASH_HOLDER)) {
                            entity.offer(Keys.LEASH_HOLDER, player);
                        }
                        player.sendMessage(Texts.of(TextColors.GREEN, "Entity Created!"));
                        return CommandResult.success();
                    }
                    return CommandResult.empty();
                }).build();
        Sponge.getCommandDispatcher().register(this, makeEntity, "makeEntity", "mkE");
    }
}
```

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeLeashData.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @Override
+    public LeashData asMutable() {
+        return new SpongeLeashData(getValue());
+    }
+
+    @Override
+    public ImmutableValue<Entity> leashHolder() {
+        return new ImmutableSpongeEntityValue<>(Keys.LEASH_HOLDER, getValue());
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeLeashData.java]: This should be stored in a field.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeLeashData.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        super(LeashData.class, value, Keys.LEASH_HOLDER);
+    }
+
+    public SpongeLeashData() {
+        this(null);
+    }
+
+    @Override
+    public LeashData copy() {
+        checkState(getValue() != null);
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeLeashData.java]: This should never have a null value.

**Comment by Aaron1011** [comment]
@Rain336; I'm currently trying to figure out how we want to handle `Value<Entity>` In the implementation. For now, you should use `SpongeEntityValue`, which does the null handling for you.

**Comment by parlough** [comment]
@gabizou I know this is kind of an old post, but I was going to implement `LeashData` until I saw you reverted this PR. What do you have in mind for a changed API?

**Comment by ryantheleach** [comment]
@Meronat My guess would be replacing Entity with a UUID?

**Pull request closed without merging**